### ISSUE TITLE
Packed geometries report a constant comp_traj, and float16 positions bias compartment counts by ~3% at R=0.5um

### ISSUE THREAD

**Issue by rutgerfick** [dmrai-lab/dmipy-sim#78]
Reported by **Marco Pizzolato**, who found it by hand-checking spin conservation on
gamma-distributed packed cylinders (minimum diameter 1 µm) and noticed the intracellular
count changed over a walk that should conserve it. Two defects compound; both reproduce on
`main` (6e7fd35).

## A. `comp_traj` is a constant for every packed geometry

`core.py:1198-1209` builds the compartment classifier only for a **permeable** single
Cylinder/Sphere; everything else falls through to a constant:

```python
# Permeable Cylinder (has _R and radius): 0=inside, 1=outside.  Permeable
# Sphere (radius, no _R): 0=inside, 1=outside.  Everything else: always 0.
...
else:
    def _get_comp_id(r):
        return jnp.int8(0)
```

So `PackedCylinders`, `PackedSpheres` and `PackedCurvedTubes` all return
`comp_traj ≡ 0` from `save_relaxation_data=True`. A user who wants per-compartment
occupancy has no option but to re-derive it from the stored positions — which walks
straight into (B).

This is gratuitous: all three geometries already implement `classify_position`
(`geometries.py:1058`, `:1155`, `curved_tube.py:55`, `:126`). The classifier should use it
when present instead of special-casing on `hasattr(geometry, 'radius')` and permeability.

## B. `simulate_trajectories` returns positions as float16 *in metres*

Positions are hard-cast at six sites (`core.py:1470, 1484, 1494, 1522, 1530, 1537`) with no
way to opt out.

The failure is not generic precision loss — it is **one-directional and therefore biased**.
A compartment is populated right up to its wall and empty beyond, so quantization can only
eject walkers, never recruit them. Measured on a uniform-in-area confined population
(400k samples, cylinder offset 12 µm into a substrate):

| cylinder radius | walkers reading OUTSIDE after an f16 round-trip |
|---|---|
| 0.5 µm | **3.144 %** |
| 1.0 µm | 1.146 % |
| 2.0 µm | 0.767 % |
| 5.0 µm | 0.297 % |

It scales as ~1/R, so it is specifically a small-axon problem. The extra-axonal population
is biased the *other* way — for R = 0.5 µm at pitch 1.2 µm, **3.30 % leak inward**, which on
an ICVF of 0.545 is +2.8 % relative. Two opposite biases on a gamma distribution weighted
toward 1 µm diameters is consistent with the 0.65 → 0.70 Marco measured.

**Root cause.** float16's smallest normal is `6.1e-5`, so *every micron-scale coordinate
expressed in metres is subnormal*, where the spacing is a flat `6e-8 m` and all of f16's
relative precision is discarded. At a 0.5 µm coordinate that is a 12.5 % quantum.

**f16 is the only non-f32 link in the chain.** The walk computes in f32; the pack stores f32
(`pack_position_arrays` defaults to f32, `encode_bridge_dst` passes it explicitly); and
`SPEC.md` §5.1 requires `positions` to be float32 with float64 permitted — f16 is not a
conformant decoded dtype. Because the pack encodes *from* the master array, an f16 master
also bakes ~60 nm of error into the bridge codec's exact-by-construction f32 endpoints:
false precision, with no size saving in the shipped `.rpk`.

## What is NOT affected

The **signal**. Phase error goes as `q·δr ≈ 1.8e5 × 6e-8 ≈ 0.01 rad`, and measured on 20k
paths at the DiSCo protocol:

```
b=  1000: S_f32=0.54984  S_f16=0.54984  |Δ|=2.8e-06
b=  3094: S_f32=0.15908  S_f16=0.15907  |Δ|=2.4e-06
b= 13192: S_f32=0.00372  S_f16=0.00371  |Δ|=1.3e-05
```

No replay or compression fidelity result is invalidated — this is a classification defect,
never a physics one, which is exactly why it survived this long.

## Fix

1. Use `geometry.classify_position` for `comp_traj` whenever the geometry provides it, so
   packed geometries report real occupancy. This is the correct fix for the original
   measurement and it is *cheaper* than storing f32 positions — int8, 1 byte/save against 12.
2. Add `storage_dtype` to `simulate_trajectories`, **defaulting to `float32`**, with
   `float16` available as an explicit opt-in. f32 is the only setting with exactly zero
   ejection, and it makes the pipeline uniformly f32. f16 halves peak RAM on large walks
   (2e7 walkers × 129 saves: 15.5 GB vs 31 GB), which is a real but explicit trade.
3. Regression tests: `comp_traj` is non-trivial for packed geometries; compartment counts are
   conserved under an impermeable walk; the f16 opt-in still round-trips.

Marco also independently wrote a `storage_dtype` patch and a κ=0 conservation test (which
preserved 128/256 intracellular walkers exactly once the dtype was raised). One note on
framing that test: for an impermeable geometry confinement is enforced by the reflection, so
a walker cannot leave *by construction* — the test validates the recorded compartment
history rather than confinement itself. That is precisely what catches (A), but it should not
be described as verifying containment.

Rescaling f16 (storing µm, or normalising by the substrate extent) recovers most of the
precision at identical memory — 3.14 % → 0.47 % / 0.36 % ejection at R = 0.5 µm — but it does
**not** apply to the returned master array, which must be in metres; handing back f32
reconstructed from scaled f16 costs f32 memory anyway. It is only ever a codec-storage trick,
and the codec is already f32. Recorded here so it is not re-proposed.

**Comment by rutgerfick** [comment]
Fixed in #79, squash-merged to `main` as 9c15d6f.

- `simulate_trajectories(storage_dtype=np.float32)` is now the default; `float16` stays available as an explicit opt-in for the peak-RAM saving.
- `PackedCylinders`/`PackedSpheres` now emit a real two-pool `comp_traj` (int8), collapsing their `0=extra, 1..N=object` labels — collapsing is required rather than cosmetic, since a raw object id overflows int8 above 127 objects.
- Seven regression tests, including the f16-subnormal mechanism in isolation and the κ=0 conservation check.

Worth recording for anyone who lands here: the reported symptom was compartment counting only. The **signal** was never affected (|ΔS| ≤ 1.3e-5, since q·δr ≈ 0.01 rad), which is why this survived — it was invisible from the signal side and needed exactly the manual conservation check Marco did.

One scope note: `PackedCylinders(permeability=None)` *ejects* intra-seeded walkers on the first step, so it models extra-axonal water only and `comp_traj ≡ 0` is correct and complete there. The defect was the permeable path. A test now pins the impermeable case so it does not get "fixed" later.

Two follow-ups split out rather than bundled into the fix: the compartment-id convention split (#80) and a pre-existing dispatch-test failure found while validating (#81).
